PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc

```diff
@@ -494,44 +475,70 @@ class GenWeightsTableProducer : public edm::global::EDProducer<edm::StreamCache<
   }
 
   void fillOnlyPSWeightTable(Counter* counter,
+                             const DynamicWeightChoiceGenInfo* genWeightChoice,
                              double genWeight,
                              const GenEventInfoProduct& genProd,
                              std::unique_ptr<nanoaod::FlatTable>& outPS) const {
-    std::size_t vectorSize =
-        (genProd.weights().size() > 2)
-            ? (keepAllPSWeights_ ? (genProd.weights().size() - 2)
-                                 : ((genProd.weights().size() == 14 || genProd.weights().size() == 46) ? 4 : 1))
-            : 1;
-    std::vector<double> wPS(vectorSize, 1);
+    std::vector<double> wPS;
     std::string psWeightDocStr;
-    if (vectorSize > 1) {
-      double nominal = genProd.weights()[1];  // Called 'Baseline' in GenLumiInfoHeader
-      if (keepAllPSWeights_) {
-        for (std::size_t i = 0; i < vectorSize; i++) {
-          wPS[i] = (genProd.weights()[i + 2]) / nominal;
-        }
-        psWeightDocStr = "All PS weights (w_var / w_nominal)";
-      } else {
-        for (std::size_t i = 6; i < 10; i++) {
-          wPS[i - 6] = (genProd.weights()[i]) / nominal;
-        }
-        psWeightDocStr =
-            "PS weights (w_var / w_nominal); [0] is ISR=0.5 FSR=1; [1] is ISR=1 "
-            "FSR=0.5; [2] is ISR=2 FSR=1; [3] is ISR=1 FSR=2 ";
-      }
-    } else {
-      psWeightDocStr = "dummy PS weight (1.0) ";
-    }
+    setPSWeightInfo(genProd.weights(), genWeightChoice, wPS, psWeightDocStr);
     outPS = std::make_unique<nanoaod::FlatTable>(wPS.size(), "PSWeight", false);
     outPS->addColumn<float>("", wPS, psWeightDocStr, lheWeightPrecision_);
 
     counter->incGenOnly(genWeight);
     counter->incPSOnly(genWeight, wPS);
   }
 
-  // create an empty counter
-  std::shared_ptr<DynamicWeightChoice> globalBeginRun(edm::Run const& iRun, edm::EventSetup const&) const override {
-    edm::Handle<LHERunInfoProduct> lheInfo;
+  void setPSWeightInfo(const std::vector<double>& genWeights,
+                       const DynamicWeightChoiceGenInfo* genWeightChoice,
+                       std::vector<double>& wPS,
+                       std::string& psWeightDocStr) const {
+    bool isRegularPSSet = (genWeights.size() == 14 || genWeights.size() == 46);
+    bool choiceSameSizeAsWeights = !keepAllPSWeights_ || (genWeights.size() - 2 == genWeightChoice->psWeightIDs.size());
+    if (!genWeightChoice->empty() && (!isRegularPSSet || choiceSameSizeAsWeights)) {
```

Comment:
This logic is really convoluted, yah... The may reason for this is one sample I checked had vaid PS weights, but the LumiHeader had several of the weight names empty. Since the logic of the code is to try and add weights based on the names in the header, if some weights names are missing, those weights are skipped. So in human readable version, what I wanted to say was:

Only fill from psWeightIDs if the WeightChoice is not empty (header exists) AND if there the same number of weights as a usual PS set then the number of names in WeightChoice must also equal the number of weights.

I put the weird logic to get that if...then... logic, but that might be unnecessarily confusing. I've been considering just skipping using the WeightChoice indices is `(genWeights.size() == 14 || genWeights.size() == 46)`, that would certainly clean up this logic a lot. I'll make that change

Reply:
I changed to your logic (I think...). So the two cases are filling using header names or set all weights as PS weights. The logic now is 
```
bool isRegularPSSet = keepAllPSWeights_ && (genWeights.size() == 14 || genWeights.size() == 46);
if(!genWeightChoice->psWeightIDs.empty() && !isRegularPSSet) 
    // Fill using header names
else 
    // Fill all weights
```
So if the header is empty OR we want to keep all the weights and the genWeights are of a usual PS size, go ahead and just fill all the weights. Hopefully this makes things cleaner and avoid problems mentioned before of lumiheaders missing weight names.

Let me know if this is better or I should change the logic a bit more!

Reply:
Thanks for the update, I find it clearer now! So to summarize the logic is now:

- keepAllPSWeights:
  - header present
    - regular weight size (14 or 46) -> fill all ([2,n])
    - irregular weight size -> fill using names (i.e. all with isr/fsr in name)
  - no header
    - fill all ([2,n])
- NOT keepAllPSWeights:
  - header present -> try to match names (+fill missing names, depending on regular/alt order)
    - found PS weights -> fill using names
    - no PS weights -> fill dummy (1)
  - no header
    - regular weight size -> fill entries 6-9 [might be wrong if alternative order, issue warning]
    - irregular weight size -> fill dummy (1)

That seems reasonable to me!

Reply:
Exactly! you wrote this out in a much better way than I could... :stuck_out_tongue_closed_eyes: